### PR TITLE
 Using EIP-1559 V2 for approve transaction flow

### LINKED ISSUE #12880: Refactor approve token flow to use new V2 functionality
Approve transaction flow should use new V2 edit / advance gas fee popups

Screenshot 2021-11-29 at 8 30 42 PM

### PR DESCRIPTION
**Pull request by jpuri**
Fixes: #12880

Explanation:  Using V2 for approve transaction flow

https://user-images.githubusercontent.com/2182307/144082672-926fdbdc-5de7-4e03-8eae-aa87d79e1a46.mov

### CODE REVIEW

File: ui/pages/confirm-approve/confirm-approve.js [hunk 1]
```diff
@@ -163,101 +181,106 @@ export default function ConfirmApprove() {
...
-              mode={EDIT_GAS_MODES.MODIFY_IN_PLACE}
-              transaction={transaction}
+    <GasFeeContextProvider transaction={transaction}>
+      <ConfirmTransactionBase
+        toAddress={toAddress}
+        identiconAddress={tokenAddress}
+        showAccountInHeader
+        title={tokensText}
+        contentComponent={
+          <TransactionModalContextProvider captureEventEnabled={false}>
```
Comment (mcmire) [inline ui/pages/confirm-approve/confirm-approve.js:191]: `captureEventEnabled={false}` — was this moved from somewhere else or is this a new thing? What does this do?

**Review by mcmire — DISMISSED** [review]
Had a question, but this looks good to me.

File: ui/pages/confirm-approve/confirm-approve.js (hunk 1 above)
Comment (jpuri) [reply ui/pages/confirm-approve/confirm-approve.js:191]: This is new route and different flow - approve token. Please check video in description.

File: ui/components/app/edit-gas-fee-button/edit-gas-fee-button.js [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    hasSimulationError,
+    estimateUsed,
+    maxFeePerGas,
+    maxPriorityFeePerGas,
+    supportsEIP1559V2,
+    transaction,
+  } = useGasFeeContext();
+  const { openModal } = useTransactionModalContext();
+
+  if (supportsEIP1559V2 && estimateUsed) {
```
Comment (darkwing) [inline ui/components/app/edit-gas-fee-button/edit-gas-fee-button.js:25]: Instead of this large `if`, could we have a `return null` at the top if the condition isn't met?

Comment (jpuri) [reply ui/components/app/edit-gas-fee-button/edit-gas-fee-button.js]: Good point, I just fixed

**Review by mcmire — DISMISSED** [review]

**Review by mcmire — DISMISSED** [review]

File: ui/components/app/edit-gas-fee-button/index.scss [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+  }
+
+  &__icon {
+    font-size: 1rem;
+  }
+
+  &__label {
+    font-size: 12px;
+    margin-right: 8px;
+  }
```
Comment (adonesky1) [inline ui/components/app/edit-gas-fee-button/index.scss:30]: [nit] maybe we should try to start staying consistent with either `rem` or `px` for these sizes? Was there a reason you used `1rem` for icon and `px` for everything else?

File: ui/components/app/edit-gas-fee-popover/edit-gas-item/edit-gas-item.js [hunk 4]
```diff
@@ -46,14 +46,18 @@ const EditGasItem = ({ priorityLevel }) => {
...
-  ) {
-    maxFeePerGas = hexWEIToDecGWEI(dappSuggestedGasFees.maxFeePerGas);
-    maxPriorityFeePerGas = hexWEIToDecGWEI(
-      dappSuggestedGasFees.maxPriorityFeePerGas,
-    );
+  } else if (priorityLevel === PRIORITY_LEVELS.DAPP_SUGGESTED) {
+    if (estimateUsed === PRIORITY_LEVELS.DAPP_SUGGESTED) {
+      maxFeePerGas = maxFeePerGasValue;
+      maxPriorityFeePerGas = maxPriorityFeePerGasValue;
+    } else if (dappSuggestedGasFees) {
```
Comment (adonesky1) [inline ui/components/app/edit-gas-fee-popover/edit-gas-item/edit-gas-item.js:53]: Could you explain why this change was made? Why do we want to select the values stored in the `gasFeeContext` over the `dappSuggestedFees` on the `transaction` object?

File: ui/components/app/edit-gas-fee-button/index.scss (hunk 3 above)
Comment (mcmire) [reply ui/components/app/edit-gas-fee-button/index.scss:30]: I had a similar question but I wasn't sure about it. Do we have conventions around when to use `rem` vs. `px`?

File: ui/pages/confirm-approve/confirm-approve-content/confirm-approve-content.component.js [hunk 5]
```diff
@@ -76,11 +78,13 @@ export default class ConfirmApproveContent extends Component {
     symbol,
     title,
     showEdit,
+    showCustomizeGasModal = false,
```
Comment (adonesky1) [inline ui/pages/confirm-approve/confirm-approve-content/confirm-approve-content.component.js:81]: having trouble following why this is necessary?

Comment (adonesky1) [reply ui/pages/confirm-approve/confirm-approve-content/confirm-approve-content.component.js:81]: I think the naming here is super confusing because `showCustomizeGasModal` is a function in `ConfirmApproveContent` and you're using the same name for a boolean here in this method on the `ConfirmApproveContent` component.

File: ui/pages/confirm-approve/confirm-approve.js [hunk 6]
```diff
@@ -163,101 +176,106 @@ export default function ConfirmApprove() {
...
-    />
+            {showCustomizeGasPopover && !supportsEIP1559V2 && (
+              <EditGasPopover
+                onClose={closeCustomizeGasPopover}
+                mode={EDIT_GAS_MODES.MODIFY_IN_PLACE}
+                transaction={transaction}
+              />
+            )}
+            <EditGasFeePopover />
+            <AdvancedGasFeePopover />
```
Comment (adonesky1) [inline ui/pages/confirm-approve/confirm-approve.js:273]: Should we wrap these like :
```
{showCustomizeGasPopover && supportsEIP1559V2 && (
             <EditGasFeePopover />
            <AdvancedGasFeePopover />
            )}
```

**Review by adonesky1 — COMMENTED** [review]
A few small questions.

File: ui/components/app/edit-gas-fee-button/index.scss (hunk 3 above)
Comment (jpuri) [reply ui/components/app/edit-gas-fee-button/index.scss:30]: No specific reason to use `rem` for icon, as I remember it was some design feedback after which i changed the icon size to `1rem`.
I fixed the file to use `px` everywhere.

File: ui/components/app/edit-gas-fee-popover/edit-gas-item/edit-gas-item.js (hunk 4 above)
Comment (jpuri) [reply ui/components/app/edit-gas-fee-popover/edit-gas-item/edit-gas-item.js:53]: In V2 implementation values from `useGasFeeContext()` are dapp suggested if estimateUsed is `dappSuggested`.

File: ui/pages/confirm-approve/confirm-approve-content/confirm-approve-content.component.js (hunk 5 above)
Comment (jpuri) [reply ui/pages/confirm-approve/confirm-approve-content/confirm-approve-content.component.js:81]: fixed

File: ui/pages/confirm-approve/confirm-approve.js (hunk 6 above)
Comment (jpuri) [reply ui/pages/confirm-approve/confirm-approve.js:273]: `EditGasFeePopover` and `AdvancedGasFeePopover` will return null if `supportsEIP1559V2` is false.

File: ui/components/app/edit-gas-fee-popover/edit-gas-item/edit-gas-item.js (hunk 4 above)
Comment (jpuri) [reply ui/components/app/edit-gas-fee-popover/edit-gas-item/edit-gas-item.js:51]: I anyways updated this to always use `dappSuggestedGasFees` from `transaction`.

File: ui/components/app/edit-gas-fee-button/index.scss (hunk 3 above)
Comment (adonesky1) [reply ui/components/app/edit-gas-fee-button/index.scss:30]: Yeah would this `label` work using one of the `mixins`?

**Review by adonesky1 — APPROVED** [review]
One small non-blocking question left. Otherwise LGTM!

File: ui/pages/confirm-approve/confirm-approve-content/confirm-approve-content.component.js [hunk 7]
```diff
@@ -76,11 +78,13 @@ export default class ConfirmApproveContent extends Component {
     symbol,
     title,
     showEdit,
+    showAdvanceGasFeeOptions = false,
```
Comment (mcmire) [inline ui/pages/confirm-approve/confirm-approve-content/confirm-approve-content.component.js:81]: `showAdvancedGasFeeOptions`?

**Review by mcmire — APPROVED** [review]
One nit as well, but otherwise looks good to me.

File: ui/components/app/edit-gas-fee-button/index.scss (hunk 3 above)
Comment (jpuri) [reply ui/components/app/edit-gas-fee-button/index.scss:30]: I expect it to

File: ui/pages/confirm-approve/confirm-approve-content/confirm-approve-content.component.js (hunk 7 above)
Comment (jpuri) [reply ui/pages/confirm-approve/confirm-approve-content/confirm-approve-content.component.js:81]: Let me fix this in following PR

Comment (mcmire) [reply ui/pages/confirm-approve/confirm-approve-content/confirm-approve-content.component.js:81]: No problem!
